src/gsi/core_disp_func.F90

```diff
@@ -0,0 +1,1055 @@
+     i1=max(ids+1,its)-its+1
+     i2=min(ide-1,ite)-its+1
+  end if
+    
+  do j = j1, j2
+     do i = i1, i2
+        if (ix_orig(i,j) .lt. ips .or. ix_orig(i,j) .gt. ipe .or. &
+             iy_orig(i,j) .lt. jps .or. iy_orig(i,j) .gt. jpe) then
+           ! outside patch:
+           needed_ij_local(i,j,1:2) = (/ix_orig(i,j), iy_orig(i,j)/)
+           num_needed_in=num_needed_in+1
+        end if
+     end do
+  end do
+
+  allocate(needed_ij_in(4,num_needed_in), needed_wgts_in(2,num_needed_in),index_in(num_needed_in))
+  offset=0
+  do j = j1, j2
+     do i = i1, i2
+        if (needed_ij_local(i,j,1) .ne. BAD_INT) then
+           offset=offset+1
+           needed_ij_in(1:4,offset) = (/i+its-1,j+jts-1,needed_ij_local(i,j,1),needed_ij_local(i,j,2)/)
+           needed_wgts_in(1:2,offset) = (/dx_n(i,j),dy_n(i,j)/)
+        end if
+     end do
+  end do
+
+  ! Call a routine that exchanges this info with all other patches:
+  ! NOTE: needed_ij_out, needed_wgts_out is allocated within exchange_needed
+
+  call exchange_needed(needed_ij_in, needed_wgts_in, num_needed_in, &
+       ips, ipe, jps, jpe, needed_ij_out, needed_wgts_out, num_needed_out, &
+       num_glob_needed, index_in)
+
+  deallocate(needed_wgts_in,stat=ierr)
+
+#ifdef TRACE_USE
+  if (trace_use) call da_trace_exit("compute_xy_orig_dm")
+#endif
+end subroutine compute_xy_orig_dm
+
+subroutine exchange_needed(needed_ij_in, needed_wgts_in, num_needed_in, &
+     dum_ips, dum_ipe, dum_jps, dum_jpe, needed_ij_out, needed_wgts_out, &
+     num_needed_out, num_glob_needed, index_in)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram: exchange_needed
+!   prgmmr: Nehrkorn
+!
+! abstract: This routine exchanges information about displacement vectors (starting and
+!           end points, and interpolation weights) between processors
+!           patch dimensions: dum_ips:dum_ipe, dum_jps:dum_jpe
+
+! program history log:
+!
+!   input argument list: 
+!      needed_ij_in:
+!       displacement vectors with endpoints in current patch (i,j stored in columns 1,2)
+!       and origin points in other patches (ix_orig, iy_orig stored in columns 3,4)
+!       NOTE: reordered upon output
+!      needed_wgts_in
+!       associated interpolation weights (x,y stored in columns 1,2)
+!       NOTE: not needed after this routine
+!      index_in:
+!       ith element points to location in re-broadcasted global array of needed_out points/values
+!       for the ith row of (reordered) needed_ij_in
+!
+!   output argument list:
+!      needed_ij_out:
+!       displacement vectors with endpoints in other patches (i,j stored in columns 1,2)
+!       and origin points in current patch (ix_orig, iy_orig stored in columns 3,4)
+!      needed_wgts_out
+!       associated interpolation weights (x,y stored in columns 1,2)
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  use mpi
```

Comment:
All module uses need to be explicitly declared:

use mpi, only:

Reply:
Done.